### DISCUSSION TITLE
Gas estimation errored with the following message (see below). The transaction execution will likely fail. Do you want to force sending? Error happened while trying to execute a function inside a smart contract

### DISCUSSION THREAD

**Discussion by alieneddie** [community/community#109757] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

Hi, i am new to coding and i have made a simple contract with chat gpt but i keep getting this error and I dont know why. Every other function works properly, its very frustrating.

Here is my contract
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
import "@openzeppelin/contracts-upgradeable/utils/math/SafeMathUpgradeable.sol";
import "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";

contract AERTokenPool is Initializable, OwnableUpgradeable, ReentrancyGuardUpgradeable {
    using SafeMathUpgradeable for uint256;
    using SafeERC20Upgradeable for IERC20Upgradeable;

    IERC20Upgradeable public aerToken;
    uint256 public constant minPoolBalance = 1 * (10 ** 18); // Minimum pool balance required
    uint256 public totalRewards; // Total accumulated rewards in the pool
    uint256 public totalStreamingDuration; // Total streaming duration of all users

    mapping(address => uint256) public streamingDuration;

    event RewardsAdded(uint256 amount);
    event RewardsClaimed(address indexed user, uint256 amount);

    function initialize(address _aerTokenAddress, address _owner) initializer public {
        __Ownable_init(_owner); // Set the initial owner
        aerToken = IERC20Upgradeable(_aerTokenAddress);
    }

    // Function to add rewards to the pool
    function addRewards(uint256 amount) external onlyOwner nonReentrant {
        totalRewards = totalRewards.add(amount);
        aerToken.safeTransferFrom(msg.sender, address(this), amount);
        emit RewardsAdded(amount);
    }

    // Function to calculate user's share of rewards based on AER holdings and streaming duration
    function calculateRewards(address user) internal view returns (uint256) {
        uint256 userBalance = aerToken.balanceOf(user);
        uint256 poolBalance = aerToken.balanceOf(address(this));
        uint256 userDuration = streamingDuration[user];

        if (poolBalance == 0 || userBalance == 0 || userDuration == 0) {
            return 0; // No rewards if either the pool balance, user balance, or streaming duration is zero
        }

        // Calculate the user's share of rewards proportionally to their balance and streaming duration
        uint256 userShare = totalRewards.mul(userBalance).mul(userDuration).div(poolBalance.mul(totalStreamingDuration));

        return userShare;
    }

    // Function to allow users to claim their share of rewards
    function claimRewards() external nonReentrant {
        uint256 userRewards = calculateRewards(msg.sender);
        require(userRewards > 0, "No rewards to claim");

        totalRewards = totalRewards.sub(userRewards);
        aerToken.safeTransfer(msg.sender, userRewards);
        emit RewardsClaimed(msg.sender, userRewards);

        // Reset the streaming duration for the user
        totalStreamingDuration = totalStreamingDuration.sub(streamingDuration[msg.sender]);
        streamingDuration[msg.sender] = 0;
    }

    // Function to update the streaming duration for a user
    function updateStreamingDuration(address user, uint256 durationInSeconds) external {
        streamingDuration[user] = streamingDuration[user].add(durationInSeconds);
        totalStreamingDuration = totalStreamingDuration.add(durationInSeconds);
    }

    // Function to check if the minimum pool balance is maintained
    function checkMinPoolBalance() external view returns (bool) {
        return aerToken.balanceOf(address(this)) >= minPoolBalance;
    }
}

**Comment by alieneddie** [comment]
This error only happens when i call the 'addRewards' function - Gas estimation errored with the following message (see below). The transaction execution will likely fail. Do you want to force sending?
Error happened while trying to execute a function inside a smart contract

**Reply by tusharr1411** [reply]
getting a saame error for my smartcontract. It is deployed but getting the same error when calling  a function. did you get any solution ?

**Comment by shhmeks** [comment]
## Gas Estimation Error - ERC20 Approval Issue

The **gas estimation error** occurs because the `addRewards()` function calls `aerToken.safeTransferFrom(msg.sender, address(this), amount)` without the user having approved the contract to spend their tokens first, causing the transaction simulation to fail during gas estimation.

You need to **approve the contract** to spend your AER tokens before calling `addRewards()` by calling the ERC20 `approve()` function with the contract address and the amount you want to transfer. The typical workflow involves first calling `aerToken.approve(contractAddress, amount)` to give permission, then calling `addRewards(amount)` to actually transfer the tokens.

Consider implementing a **two-step process** where users first approve the contract, then call `addRewards()`, or add a helper function that checks current allowance and provides clear error messages.
